themes/default/content/blog/resource-methods-for-pulumi-packages/index.md

```diff
@@ -0,0 +1,609 @@
+
+<!--more-->
+
+When authoring component resources, it's often useful to provide additional functionality through methods on the component. For example, the `Cluster` component in the [`eks`]({{<relref "/docs/reference/pkg/eks">}}) package has a [`getKubeconfig`](https://github.com/pulumi/pulumi-eks/blob/700d73e961976e58762cb9c723ad2fa838052f46/nodejs/eks/cluster.ts#L1482) method that can be used to generate a kubeconfig for authentication with the cluster that does not use the default AWS credential provider chain, but is instead scoped based on the passed-in arguments. Until now, that method has only been available from JavaScript/TypeScript (the language the `Cluster` component was written in). With the new support for resource methods for Pulumi Packages, we can make this method available to all the Pulumi languages, which is exactly what we've done in [#628](https://github.com/pulumi/pulumi-eks/pull/628).
+
+{{< chooser language "typescript,python,csharp,go" >}}
+
+{{% choosable language typescript %}}
+
+```typescript
+const cluster = new eks.Cluster("mycluster");
+const kubeconfig = cluster.getKubeconfig(...);
+```
+
+{{% /choosable %}}
+{{% choosable language python %}}
+
+```python
+cluster = eks.Cluster("mycluster")
+kubeconfig = cluster.get_kubeconfig(...)
+```
+
+{{% /choosable %}}
+{{% choosable language csharp %}}
+
+```csharp
+var cluster = new Cluster("mycluster");
+var kubeconfig = cluster.GetKubeconfig(...);
+```
+
+{{% /choosable %}}
+
+{{% choosable language go %}}
+
+```go
+cluster, err := eks.NewCluster(ctx, "mycluster", nil)
+if err != nil {
+    return err
+}
+kubeconfig, err := cluster.GetKubeconfig(ctx, &eks.ClusterGetKubeconfigArgs{...})
+if err != nil {
+    return err
+}
+```
+
+{{% /choosable %}}
+
+{{% /chooser %}}
+
+This post will show how you can provide resource methods from your [component packages]({{<relref "/docs/guides/pulumi-packages">}}).
+
+## Authoring Methods
+
+It's always been possible to provide module-level functions from Pulumi Packages. For example, the [`aws`]({{<relref "/docs/reference/pkg/aws">}}) package provides many module-level functions, such as the [`getAmi`]({{<relref "docs/reference/pkg/aws/ec2/getami">}}) function which can be used to get the ID of an existing Amazon Machine Image (AMI). Functions are declared in the package's [schema]({{<relref "/docs/guides/pulumi-packages/schema">}}) and their functionality is implemented in the provider through the `Invoke` RPC.
+
+Methods are authored in a similar manner to functions. Methods are declared in the schema and implemented in the provider's `Call` RPC (similar to `Invoke`).
+
+Let's walk through an example. We'll author a `Message` component which accepts a message as input and provides a `getMessage` method that accepts a recipient name and returns the message customized for the recipient. To get started authoring a component package, refer to the [package]({{<relref "/docs/guides/pulumi-packages">}}) documentation.
```

Comment:
Patching the relref and breaking up the sentence. Some readers will have trouble comprehending the longer sentence, so this reword makes it easier for everyone to follow the conversation.

```suggestion
Let's walk through an example. We'll author a `Message` component that accepts a message as input. The component then provides a `getMessage` method that accepts a recipient name and returns the message customized for the recipient. To get started authoring a component package, refer to the [package]({{< relref "/docs/guides/pulumi-packages" >}}) documentation.
```